src/librustc_codegen_llvm/sir.rs

```diff
@@ -32,6 +32,14 @@ pub fn write_sir<'tcx>(tcx: TyCtxt<'tcx>, sir_llvm_module: &mut ModuleLlvm) {
     let mut buf = Vec::new();
     let mut encoder = ykpack::Encoder::from(&mut buf);
 
+    // First we serialise the types which will be referenced in the body packs that will follow.
+    // The serialisation order matters here, as the load order (in the runtime) corresponds with
+    // the type indices, hence use of `IndexMap` for insertion order.
+    let types =
+        tcx.sir_types.borrow_mut().map.drain(..).map(|(k, _)| k).collect::<Vec<ykpack::Ty>>();
```

Comment:
Is `drain` the right thing here? I was perhaps expecting `into_iter`.

Reply:
Wouldn't we have to consume the `tcx `to use `into_iter`? Will try it.

Reply:
I guess the question is more: do we ever need access to `sir_types` after this point? If not, it doesn't really matter which one we do.

Reply:
Ah, if you try to use `into_iter` then you end up trying to move out of a `Rc` which isn't allowed.

I figured `drain` was better than putting another inner-mutable wrapper around the innards of the `Rc`. We won't need the types again once they are serialised.

Reply:
> We won't need the types again once they are serialised.

OK, then it doesn't matter. Sorry for the noise!